etcdserver/api/v3rpc/watch.go

```diff
@@ -117,10 +123,30 @@ func (sws *serverWatchStream) recvLoop() error {
 				id = sws.watchStream.Watch(creq.Key, creq.RangeEnd, rev)
 			}
 			sws.ctrlStream <- &pb.WatchResponse{
-				Header:   sws.newResponseHeader(wsrev),
-				WatchId:  int64(id),
-				Created:  true,
-				Canceled: futureRev,
+				Header:        sws.newResponseHeader(wsrev),
+				WatchId:       int64(id),
+				Created:       true,
+				Canceled:      futureRev,
+				StartRevision: rev,
+			}
+
+			if uv.CreateRequest.ProgressReport {
+				go func() {
+					for {
+						select {
+						case <-sws.reportDone:
+							close(sws.reportStream)
+							return
+						case <-time.After(time.Duration(uv.CreateRequest.ReportInterval) * time.Second):
```

Comment:
Potential race: events posted to watcher channel for rev1, watcher revision increased to rev2, timer fires and progress report is sent with rev2, sendLoop then receives events from watcher channel and sends events for rev1 after reporting progress at rev2. 

This should at least be in the sendLoop.